### DISCUSSION TITLE
zoom doesnt work on orthographic camera, while OpenGL projection is 3D

### DISCUSSION THREAD

**Discussion by 10100010010001001010** [axmolengine/axmol#2258] [Q&A]
```
    camera = Camera::createOrthographic(1920, 1080, -1024, 1024);
    camera->setCameraFlag(CameraFlag::USER1);
    addChild(camera);
```
this is how I create my camera, and it does not apply the specified zoom, until I do this thing in my init scene function: 
`_director->setProjection(ax::Director::Projection::_2D);`

I want to know if this is how I should do it, or if this is a bug in the engine code ? (or maybe I made a mistake somewhere )
As for the standard camera, when you increase the zoom, the culling starts to work incorrectly, and standard sprites are cut off before they disappear from the screen, while spine elements are displayed correctly.

**Comment by rh101** [comment]
Are you able to provide a minimal test case, screenshot or video of what it is you are seeing, and what it is you expect to see?

> As for the standard camera, when you increase the zoom, the cooling starts to work incorrectly, and standard sprites are cut off before they disappear from the screen, while spine elements are displayed correctly.

Do you mean culling?

**Reply by 10100010010001001010** [reply]
yes, I mean culling. Look:
![11](https://github.com/user-attachments/assets/cb541021-e330-4b20-a92d-23ee3dc5edd0)
![22](https://github.com/user-attachments/assets/eb5ab932-0ee2-4849-8cd0-57bfce856c8b)
This is the default camera with `setZoom(0.8)`. In the first image everything is fine, but as soon as you move the camera a little to the left, the image stops being displayed, although it is clear from its hitbox that the camera still sees it. With standard zoom everything is displayed as it should

if I try to use an orthographic camera, instead default, the zoom simply does not work, although I checked its value while the program was running and it was the same as I had set. As soon as I set Projection::_2D, everything starts working correctly. Including the display of this sprite with orthographic camera

**Reply by DelinWorks** [reply]
yeah this is expected with zoom that is less than 1.0, I noticed it in axmol a long time ago and I think I fixed it for tilemaps but I don't know about sprites, the culling logic was flawed

**Reply by 10100010010001001010** [reply]
@rh101, what do you think ?

**Reply by rh101** [reply]
> @rh101, what do you think ?

Sorry, I don't have anything to add to this, since it's not an area I'm familiar with.  I still strongly suggest that you create a sample project and post it somewhere, and perhaps someone can help when they can actually see the issue.

**Comment by DelinWorks** [comment]
have you tried calling setZoom on the camera object?

**Reply by 10100010010001001010** [reply]
this is literally what I do next in my code, I just described how I create the camera, just in case

**Reply by DelinWorks** [reply]
call applyZoom

**Reply by 10100010010001001010** [reply]
doesn't work. And besides, setZoom itself calls this method. I described the problem more accurately above.

**Comment by TedGriggs** [comment]
I'm not sure if this is quite the same case, but I also found that setZoom() did not work with an orthographic camera except when calling setZoom() immediately after creating the camera.  So, once scene was being rendered calling setZoom() did nothing.  In looking at the code in Camera.cpp , I added " _viewProjectionDirty = true;" in setZoom() immediately before it calls applyZoom().  This fixed it for me.

**Comment by TedGriggs** [comment]
To be clear, here's the modified function:


void Camera::setZoom(float factor)
{
    _zoomFactor = factor;
    _viewProjectionDirty = true;
    applyZoom();
}

**Comment by TedGriggs** [comment]
Related to this -- I thank the OP with regards to his/her comment about calling: _director->setProjection(ax::Director::Projection::_2D);

It makes me curious why this is set in the director and not a property of the camera?  Consider having the projection as 3D, you create an ortho camera and call applyZoom().  applyZoom then turns it into a 3D camera, based on the director settings.  I'm probably missing something, but this seems clearly wrong.
